File: src/main/java/org/example/kalkulationsprogramm/domain/Mitarbeiter.java
package org.example.kalkulationsprogramm.domain;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.OneToMany;
import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
@Entity
public class Mitarbeiter {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String vorname;

    @Column(nullable = false)
    private String nachname;

    private String strasse;
    private String plz;
    private String ort;

    @Column
    private String email;

    @Column
    private String telefon; // Mobiltelefon

    @Column
    private String festnetz; // Festnetznummer

    @Enumerated(EnumType.STRING)
    @Column
    private Qualifikation qualifikation;

    @Column
    private LocalDate geburtstag;

    @Column
    private LocalDate eintrittsdatum;

    @Column(nullable = false)
    private Boolean aktiv = true;

    @Column(precision = 10, scale = 2)
    private BigDecimal stundenlohn;

    @Column
    private Integer jahresUrlaub;

    /**
     * Resturlaub aus dem Vorjahr (in Tagen).
     * Verfällt am 1. Februar des aktuellen Jahres.
     */
    @Column
    private Integer resturlaubVorjahr;

    /**
     * Manuelle Urlaubskorrektur (in Tagen).
     * Positiv = zusätzliche Tage, Negativ = weniger Tage.
     */
    @Column
    private Integer urlaubsKorrektur;

    @Column(unique = true)
    private String loginToken;

    // N:M Beziehung - Mitarbeiter kann mehreren Abteilungen angehören
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(name = "mitarbeiter_abteilung", joinColumns = @JoinColumn(name = "mitarbeiter_id"), inverseJoinColumns = @JoinColumn(name = "abteilung_id"))
    private Set<Abteilung> abteilungen = new HashSet<>();

    // N:M – EN-1090-Rollen (WPK-Leiter, Schweißaufsicht, etc.)
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(name = "mitarbeiter_en1090_rolle", joinColumns = @JoinColumn(name = "mitarbeiter_id"), inverseJoinColumns = @JoinColumn(name = "rolle_id"))
    private Set<En1090Rolle> en1090Rollen = new HashSet<>();

    @OneToMany(mappedBy = "mitarbeiter", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<MitarbeiterQualifikation> qualifikationen = new ArrayList<>();

    @OneToMany(mappedBy = "mitarbeiter", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<MitarbeiterDokument> dokumente = new ArrayList<>();

    @OneToMany(mappedBy = "mitarbeiter", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<MitarbeiterNotiz> notizen = new ArrayList<>();

    @OneToMany(mappedBy = "mitarbeiter", cascade = CascadeType.ALL)
    private List<Lohnabrechnung> lohnabrechnungen = new ArrayList<>();
}

File: src/main/kotlin/org/example/kalkulationsprogramm/domain/Mitarbeiter.kt
package org.example.kalkulationsprogramm.domain

import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.JoinTable
import jakarta.persistence.ManyToMany
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import java.math.BigDecimal
import java.time.LocalDate

@Entity
open class Mitarbeiter {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    open var id: Long? = null

    @Column(nullable = false)
    open var vorname: String? = null

    @Column(nullable = false)
    open var nachname: String? = null

    open var strasse: String? = null
    open var plz: String? = null
    open var ort: String? = null

    @Column
    open var email: String? = null

    @Column
    open var telefon: String? = null

    @Column
    open var festnetz: String? = null

    @Enumerated(EnumType.STRING)
    @Column
    open var qualifikation: Qualifikation? = null

    @Column
    open var geburtstag: LocalDate? = null

    @Column
    open var eintrittsdatum: LocalDate? = null

    @Column(nullable = false)
    open var aktiv: Boolean? = true

    @Column(precision = 10, scale = 2)
    open var stundenlohn: BigDecimal? = null

    @Enumerated(EnumType.STRING)
    @Column(name = "beschaeftigungsart", nullable = false)
    open var beschaeftigungsart: Beschaeftigungsart? = Beschaeftigungsart.REGULAER

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "krankenkasse_id")
    open var krankenkasse: Krankenkasse? = null

    @Column(nullable = false)
    open var kinderlos: Boolean? = false

    @Column(name = "ist_geschaeftsfuehrer", nullable = false)
    open var istGeschaeftsfuehrer: Boolean? = false

    @Column(name = "kalkulatorischer_lohn_monat", precision = 12, scale = 2)
    open var kalkulatorischerLohnMonat: BigDecimal? = null

    @Column(name = "geldwert_vorteil_monat", precision = 12, scale = 2)
    open var geldwertVorteilMonat: BigDecimal? = null

    @Column
    open var jahresUrlaub: Int? = null

    @Column
    open var resturlaubVorjahr: Int? = null

    @Column
    open var urlaubsKorrektur: Int? = null

    @Column(unique = true)
    open var loginToken: String? = null

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
        name = "mitarbeiter_abteilung",
        joinColumns = [JoinColumn(name = "mitarbeiter_id")],
        inverseJoinColumns = [JoinColumn(name = "abteilung_id")],
    )
    open var abteilungen: MutableSet<Abteilung> = HashSet()

    @OneToMany(mappedBy = "mitarbeiter", cascade = [CascadeType.ALL], orphanRemoval = true)
    open var dokumente: MutableList<MitarbeiterDokument> = ArrayList()

    @OneToMany(mappedBy = "mitarbeiter", cascade = [CascadeType.ALL], orphanRemoval = true)
    open var notizen: MutableList<MitarbeiterNotiz> = ArrayList()

    @OneToMany(mappedBy = "mitarbeiter", cascade = [CascadeType.ALL])
    open var lohnabrechnungen: MutableList<Lohnabrechnung> = ArrayList()
}
